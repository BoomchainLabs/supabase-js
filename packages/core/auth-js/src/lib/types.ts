import { AuthError } from './errors'

/** One of the providers supported by GoTrue. */
export type Provider =
  | 'apple'
  | 'azure'
  | 'bitbucket'
  | 'discord'
  | 'facebook'
  | 'github'
  | 'gitlab'
  | 'google'
  | 'keycloak'
  | 'linkedin'
  | 'notion'
  | 'slack'
  | 'spotify'
  | 'twitch'
  | 'twitter'
  | 'workos'

export type AuthChangeEvent =
  | 'PASSWORD_RECOVERY'
  | 'SIGNED_IN'
  | 'SIGNED_OUT'
  | 'TOKEN_REFRESHED'
  | 'USER_UPDATED'
  | 'USER_DELETED'

export type AuthResponse =
  | {
      data: {
        user: User | null
        session: Session | null
      }
      error: null
    }
  | {
      data: {
        user: null
        session: null
      }
      error: AuthError
    }

export type OAuthResponse =
  | {
      data: {
        provider: Provider
        url: string
      }
      error: null
    }
  | {
      data: {
        provider: Provider
        url: null
      }
      error: AuthError
    }

export type UserResponse =
  | {
      data: {
        user: User
      }
      error: null
    }
  | {
      data: {
        user: null
      }
      error: AuthError
    }

export interface Session {
  provider_token?: string | null
  /**
   * The access token jwt. It is recommended to set the JWT_EXPIRY to a shorter expiry value.
   */
  access_token: string
  /**
   * A one-time used refresh token that never expires.
   */
  refresh_token: string
  /**
   * The number of seconds until the token expires (since it was issued). Returned when a login is confirmed.
   */
  expires_in: number
  /**
   * A timestamp of when the token will expire. Returned when a login is confirmed.
   */
  expires_at?: number
  token_type: string
  user: User
}

export interface UserIdentity {
  id: string
  user_id: string
  identity_data: {
    [key: string]: any
  }
  provider: string
  created_at: string
  last_sign_in_at: string
  updated_at?: string
}

export interface User {
  id: string
  app_metadata: {
    provider?: string
    [key: string]: any
  }
  user_metadata: {
    [key: string]: any
  }
  aud: string
  confirmation_sent_at?: string
  recovery_sent_at?: string
  email_change_sent_at?: string
  new_email?: string
  invited_at?: string
  action_link?: string
  email?: string
  phone?: string
  created_at: string
  confirmed_at?: string
  email_confirmed_at?: string
  phone_confirmed_at?: string
  last_sign_in_at?: string
  role?: string
  updated_at?: string
  identities?: UserIdentity[]
}

export interface UserAttributes {
  /**
   * The user's email.
   */
  email?: string

  /**
   * The user's phone.
   */
  phone?: string

  /**
   * The user's password.
   */
  password?: string

  /**
   * An email change token.
   */
  email_change_token?: string

  /**
   * A custom data object for user_metadata that a user can modify. Can be any JSON.
   */
  data?: object
}

export interface AdminUserAttributes extends UserAttributes {
  /**
   * A custom data object for user_metadata.
   *
   * Can be any JSON.
   *
   * Only a service role can modify.
   *
   * Note: When using the GoTrueAdminApi and wanting to modify a user's user_metadata,
   * this attribute is used instead of UserAttributes data.
   *
   */
  user_metadata?: object

  /**
   * A custom data object for app_metadata that.
   *
   * Only a service role can modify.
   *
   * Can be any JSON that includes app-specific info, such as identity providers, roles, and other
   * access control information.
   */
  app_metadata?: object

  /**
   * Sets if a user has confirmed their email address.
   *
   * Only a service role can modify.
   */
  email_confirm?: boolean

  /**
   * Sets if a user has confirmed their phone number.
   *
   * Only a service role can modify.
   */
  phone_confirm?: boolean
}

export interface Subscription {
  /**
   * The subscriber UUID. This will be set by the client.
   */
  id: string
  /**
   * The function to call every time there is an event. eg: (eventName) => {}
   */
  callback: (event: AuthChangeEvent, session: Session | null) => void
  /**
   * Call this to remove the listener.
   */
  unsubscribe: () => void
}

export interface UserCredentials {
  /** The user's email address. */
  email?: string
  /** The user's phone number. */
  phone?: string
  /** The user's password. */
  password?: string
  refreshToken?: string
  /** The name of the provider. */
  provider?: Provider
  oidc?: OpenIDConnectCredentials
}
export type SignInWithPasswordCredentials =
  | {
      /** The user's email address. */
      email: string
      /** The user's password. */
      password: string
      options?: SignInWithPasswordOptions
    }
  | {
      /** The user's phone number. */
      phone: string
      /** The user's password. */
      password: string
      options?: SignInWithPasswordOptions
    }
export interface SignInWithPasswordOptions {
  captchaToken?: string
}
export type SignInWithPasswordlessCredentials =
  | {
      /** The user's email address. */
      email: string
      options?: SignInWithPasswordlessOptions
    }
  | {
      /** The user's phone number. */
      phone: string
      options?: SignInWithPasswordlessOptions
    }
export interface SignInWithPasswordlessOptions {
  /** The redirect url embedded in the email link */
  emailRedirectTo?: string
  shouldCreateUser?: boolean
  captchaToken?: string
}
export type SignInWithOAuthCredentials = {
  /** One of the providers supported by GoTrue. */
  provider: Provider
  options?: SignInWithOAuthOptions
}
export interface SignInWithOAuthOptions {
  /** A URL to send the user to after they are confirmed (OAuth logins only). */
  redirectTo?: string
  /** A space-separated list of scopes granted to the OAuth application. */
  scopes?: string
  /** An object of query params */
  queryParams?: { [key: string]: string }
}

export type VerifyOtpParams = VerifyMobileOtpParams | VerifyEmailOtpParams
export interface VerifyMobileOtpParams {
  /** The user's email address. */
  email?: undefined
  /** The user's phone number. */
  phone: string
  /** The user's password. */
  token: string
  /** The user's verification type. */
  type: MobileOtpType
}
export interface VerifyEmailOtpParams {
  email: string
  phone?: undefined
  token: string
  type: EmailOtpType
}

export type MobileOtpType = 'sms' | 'phone_change'
export type EmailOtpType = 'signup' | 'invite' | 'magiclink' | 'recovery' | 'email_change'

/** The link type */
export type GenerateLinkTypes =
  | 'signup'
  | 'invite'
  | 'magiclink'
  | 'recovery'
  | 'email_change_current'
  | 'email_change_new'

export interface OpenIDConnectCredentials {
  id_token: string
  nonce: string
  provider?: Provider
  client_id?: string
  issuer?: string
}

type AnyFunction = (...args: any[]) => any
type MaybePromisify<T> = T | Promise<T>

type PromisifyMethods<T> = {
  [K in keyof T]: T[K] extends AnyFunction
    ? (...args: Parameters<T[K]>) => MaybePromisify<ReturnType<T[K]>>
    : T[K]
}

export type SupportedStorage = PromisifyMethods<Pick<Storage, 'getItem' | 'setItem' | 'removeItem'>>

export type CallRefreshTokenResult =
  | {
      session: Session
      error: null
    }
  | {
      session: null
      error: AuthError
    }
