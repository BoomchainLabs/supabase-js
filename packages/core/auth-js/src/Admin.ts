import { AdminConfig } from './Admin.types'

export default class Admin {
  url: string
  headers: AdminConfig['headers'] = {}

  /**
   * Creates a GoTrue instance for admin interactions.
   *
   * @param url  URL of the GoTrue instance.
   * @param headers  Custom headers.
   */
  constructor(url: string, options?: AdminConfig) {
    this.url = url
    if (options?.headers) {
      this.headers = { ...this.headers, ...options.headers }
    }
  }
}
