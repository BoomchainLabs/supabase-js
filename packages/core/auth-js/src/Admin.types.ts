export interface AdminConfig {
  headers?: {
    [key: string]: string
  }
}
