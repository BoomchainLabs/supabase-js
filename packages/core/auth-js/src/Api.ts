import { get, post, put } from './lib/fetch'
import { Provider, UserAttributes } from './lib/types'

export default class Api {
  url: string
  headers: {
    [key: string]: string
  }

  constructor({ url = '', headers = {} }: any) {
    this.url = url
    this.headers = headers
  }

  /**
   * Creates a new user using their email address.
   * @param email The email address of the user.
   * @param password The password of the user.
   */
  async signUpWithEmail(email: string, password: string) {
    try {
      let data: any = await post(
        `${this.url}/signup`,
        { email, password },
        { headers: this.headers }
      )
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Logs in an existing using their email address.
   * @param email The email address of the user.
   * @param password The password of the user.
   */
  async signInWithEmail(email: string, password: string) {
    try {
      let data: any = await post(
        `${this.url}/token?grant_type=password`,
        { email, password },
        { headers: this.headers }
      )
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Sends a reset request to an email address.
   * @param email The email address of the user.
   */
  async resetPasswordForEmail(email: string) {
    try {
      let data: any = await post(`${this.url}/forgotPassword`, { email }, { headers: this.headers })
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Removes a logged-in session.
   * @param jwt A valid, logged-in JWT.
   */
  async signOut(jwt: string) {
    try {
      let headers = { ...this.headers }
      headers['Authorization'] = `Bearer ${jwt}`
      let data = await post(`${this.url}/logout`, {}, { headers, noResolveJson: true })
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Generates the relevant login URL for a third-party provider.
   * @param provider One of the providers supported by GoTrue.
   */
  getUrlForProvider(provider: Provider) {
    return `${this.url}/authorize?provider=${provider}`
  }

  /**
   * Gets the user details.
   * @param jwt A valid, logged-in JWT.
   */
  async getUser(jwt: string) {
    try {
      let headers = { ...this.headers }
      headers['Authorization'] = `Bearer ${jwt}`
      let data: any = await get(`${this.url}/user`, { headers })
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Updates the user data.
   * @param jwt A valid, logged-in JWT.
   * @param attributes The data you want to update.
   */
  async updateUser(jwt: string, attributes: UserAttributes) {
    try {
      let headers = { ...this.headers }
      headers['Authorization'] = `Bearer ${jwt}`
      let data: any = await put(`${this.url}/user`, attributes, { headers })
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }

  /**
   * Generates a new JWT.
   * @param refreshToken A valid refresh token that was returned on login.
   */
  async refreshAccessToken(refreshToken: string) {
    try {
      let data: any = await post(
        `${this.url}/token?grant_type=refresh_token`,
        { refresh_token: refreshToken },
        { headers: this.headers }
      )
      return { data, error: null }
    } catch (error) {
      return { data: null, error: error.message }
    }
  }
}
