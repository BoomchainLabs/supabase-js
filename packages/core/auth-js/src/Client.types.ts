export interface ClientConfig {
  headers?: {
    [key: string]: string
  }
}
