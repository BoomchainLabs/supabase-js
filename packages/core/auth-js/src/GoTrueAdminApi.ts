import { Fetch, _request, _userResponse } from './lib/fetch'
import { resolveFetch } from './lib/helpers'
import { AdminUserAttributes, Session, User, UserResponse } from './lib/types'
import { AuthError, isAuthError } from './lib/errors'

export default class GoTrueAdminApi {
  protected url: string
  protected headers: {
    [key: string]: string
  }
  protected fetch: Fetch

  constructor({
    url = '',
    headers = {},
    fetch,
  }: {
    url: string
    headers?: {
      [key: string]: string
    }
    fetch?: Fetch
  }) {
    this.url = url
    this.headers = headers
    this.fetch = resolveFetch(fetch)
  }

  /**
   * Removes a logged-in session.
   * @param jwt A valid, logged-in JWT.
   */
  async signOut(jwt: string): Promise<{ error: AuthError | null }> {
    try {
      await _request(this.fetch, 'POST', `${this.url}/logout`, {
        jwt,
        noResolveJson: true,
      })
      return { error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { error }
      }

      throw error
    }
  }

  /**
   * Sends an invite link to an email address.
   * @param email The email address of the user.
   * @param options.redirectTo A URL or mobile address to send the user to after they are confirmed.
   * @param options.data Optional user metadata
   */
  async inviteUserByEmail(
    email: string,
    options: {
      redirectTo?: string
      data?: object
    } = {}
  ): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'POST', `${this.url}/invite`, {
        body: { email, data: options.data },
        headers: this.headers,
        redirectTo: options.redirectTo,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Generates links to be sent via email or other.
   * @param type The link type ("signup" or "magiclink" or "recovery" or "invite").
   * @param email The user's email.
   * @param options.password User password. For signup only.
   * @param options.data Optional user metadata. For signup only.
   * @param options.redirectTo The link type ("signup" or "magiclink" or "recovery" or "invite").
   */
  async generateLink(
    type:
      | 'signup'
      | 'magiclink'
      | 'recovery'
      | 'invite'
      | 'email_change_current'
      | 'email_change_new',
    email: string,
    options: {
      password?: string
      data?: object
      redirectTo?: string
    } = {}
  ): Promise<
    | {
        data: User
        error: null
      }
    | {
        data: Session
        error: null
      }
    | { data: null; error: AuthError }
  > {
    try {
      return await _request(this.fetch, 'POST', `${this.url}/admin/generate_link`, {
        body: {
          type,
          email,
          password: options.password,
          data: options.data,
          redirect_to: options.redirectTo,
        },
        headers: this.headers,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: null, error }
      }
      throw error
    }
  }

  // User Admin API
  /**
   * Creates a new user.
   *
   * @param attributes The data you want to create the user with.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async createUser(attributes: AdminUserAttributes): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'POST', `${this.url}/admin/users`, {
        body: attributes,
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Get a list of users.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async listUsers(): Promise<
    { data: { users: User[] }; error: null } | { data: { users: [] }; error: AuthError }
  > {
    try {
      const { data, error } = await _request(this.fetch, 'GET', `${this.url}/admin/users`, {
        headers: this.headers,
      })
      if (error) throw error
      return { data: { ...data }, error: null }
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { users: [] }, error }
      }
      throw error
    }
  }

  /**
   * Get user by id.
   *
   * @param uid The user's unique identifier
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async getUserById(uid: string): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'GET', `${this.url}/admin/users/${uid}`, {
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Updates the user data.
   *
   * @param attributes The data you want to update.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async updateUserById(uid: string, attributes: AdminUserAttributes): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'PUT', `${this.url}/admin/users/${uid}`, {
        body: attributes,
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }

  /**
   * Delete a user. Requires a `service_role` key.
   *
   * @param uid The user uid you want to remove.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async deleteUser(uid: string): Promise<UserResponse> {
    try {
      return await _request(this.fetch, 'DELETE', `${this.url}/admin/users/${uid}`, {
        headers: this.headers,
        xform: _userResponse,
      })
    } catch (error) {
      if (isAuthError(error)) {
        return { data: { user: null }, error }
      }

      throw error
    }
  }
}
