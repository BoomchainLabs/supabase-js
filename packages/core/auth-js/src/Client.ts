import Api from './Api'
import { isBrowser, getParameterByName, uuid } from './lib/helpers'
import { GOTRUE_URL, DEFAULT_HEADERS, STORAGE_KEY } from './lib/constants'
import { Session, User, UserAttributes, Provider, Subscription, AuthChangeEvent } from './lib/types'

const DEFAULT_OPTIONS = {
  url: GOTRUE_URL,
  autoRefreshToken: true,
  persistSession: true,
  detectSessionInUrl: true,
  headers: DEFAULT_HEADERS,
}
export default class Client {
  api: Api
  currentUser: User | null
  currentSession?: Session | null
  autoRefreshToken: boolean
  persistSession: boolean
  stateChangeEmmitters: Map<string, Subscription> = new Map()

  /**
   * Create a new client for use in the browser.
   * @param options.url The URL of the GoTrue server.
   * @param options.headers Any additional headers to send to the GoTrue server.
   * @param options.detectSessionInUrl Set to "true" if you want to automatically detects OAuth grants in the URL and signs in the user.
   * @param options.autoRefreshToken Set to "true" if you want to automatically refresh the token before expiring.
   * @param options.persistSession Set to "true" if you want to automatically save the user session into local storage.
   */
  constructor(options: {
    url?: string
    headers?: { [key: string]: string }
    detectSessionInUrl?: boolean
    autoRefreshToken?: boolean
    persistSession?: boolean
  }) {
    const settings = { ...DEFAULT_OPTIONS, ...options }
    this.currentUser = null
    this.currentSession = null
    this.autoRefreshToken = settings.autoRefreshToken
    this.persistSession = settings.persistSession
    this.api = new Api({ url: settings.url, headers: settings.headers })
    this._recoverSession()

    // Handle the OAuth redirect
    try {
      if (settings.detectSessionInUrl && isBrowser() && !!getParameterByName('access_token')) {
        this.getSessionFromUrl({ storeSession: true })
      }
    } catch (error) {
      console.log('Error getting session from URL.')
    }
  }

  /**
   * Creates a new user.
   * @param credentials The user login details.
   * @param credentials.email The user's email address.
   * @param credentials.password The user's password.
   */
  async signUp(credentials: { email: string; password: string }) {
    try {
      this._removeSession()

      let { data, error }: any = await this.api.signUpWithEmail(
        credentials.email,
        credentials.password
      )
      if (error) throw new Error(error)

      if (data?.user?.confirmed_at) {
        this._saveSession(data)
        this._notifyAllSubscribers(AuthChangeEvent.SIGNED_IN)
      }

      return { data, error: null }
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Log in an existing user, or login via a third-party provider.
   * @param credentials The user login details.
   * @param credentials.email The user's email address.
   * @param credentials.password The user's password.
   * @param credentials.provider One of the providers supported by GoTrue.
   */
  async signIn(credentials: { email?: string; password?: string; provider?: Provider }) {
    try {
      this._removeSession()
      let { email, password, provider } = credentials

      if (email && password) return this._handeEmailSignIn(email, password)
      if (provider) return this._handeProviderSignIn(provider)
      else throw new Error(`You must provide either an email or a third-party provider.`)
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Returns the user data, if there is a logged in user.
   */
  async user() {
    try {
      if (!this.currentSession?.access_token) throw new Error('Not logged in.')

      let { data, error }: any = await this.api.getUser(this.currentSession.access_token)
      if (error) throw new Error(error)

      this.currentUser = data
      return { data: this.currentUser, error: null }
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Updates user data, if there is a logged in user.
   */
  async update(attributes: UserAttributes) {
    try {
      if (!this.currentSession?.access_token) throw new Error('Not logged in.')

      let { data, error }: any = await this.api.updateUser(
        this.currentSession.access_token,
        attributes
      )
      if (error) throw new Error(error)

      this.currentUser = data
      this._notifyAllSubscribers(AuthChangeEvent.USER_UPDATED)

      return { data: this.currentUser, error: null }
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Gets the session data from a URL string
   * @param options.storeSession Optionally store the session in the browser
   */
  async getSessionFromUrl(options?: { storeSession?: boolean }) {
    try {
      if (!isBrowser()) throw new Error('No browser detected.')

      const error_description = getParameterByName('error_description')
      if (error_description) throw new Error(error_description)

      const access_token = getParameterByName('access_token')
      const expires_in = getParameterByName('expires_in')
      const refresh_token = getParameterByName('refresh_token')
      const token_type = getParameterByName('token_type')
      if (!access_token) throw new Error('No access_token detected.')
      if (!expires_in) throw new Error('No expires_in detected.')
      if (!refresh_token) throw new Error('No refresh_token detected.')
      if (!token_type) throw new Error('No token_type detected.')

      let { data: user, error }: any = await this.api.getUser(access_token)
      if (error) throw new Error(error)

      const session: Session = {
        access_token,
        expires_in: parseInt(expires_in),
        refresh_token,
        token_type,
        user,
      }
      if (options?.storeSession) {
        this._saveSession(session)
        this._notifyAllSubscribers(AuthChangeEvent.SIGNED_IN)
      }

      return { data: session, error: null }
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Signs out the current user, if there is a logged in user.
   */
  async signOut() {
    try {
      if (this.currentSession) {
        await this.api.signOut(this.currentSession.access_token)
      }
      this._removeSession()
      this._notifyAllSubscribers(AuthChangeEvent.SIGNED_OUT)
      return true
    } catch (error) {
      return { data: null, error }
    }
  }

  /**
   * Receive a notification every time an auth event happens.
   * @returns {Subscription} A subscription object which can be used to unsubcribe itself.
   */
  onAuthStateChange(callback: Function) {
    try {
      const id: string = uuid()
      let self = this
      const subscription: Subscription = {
        id,
        callback,
        unsubscribe: () => self.stateChangeEmmitters.delete(id),
      }
      this.stateChangeEmmitters.set(id, subscription)
      return { data: this.stateChangeEmmitters.get(id), error: null }
    } catch (error) {
      return { data: null, error }
    }
  }

  private async _handeEmailSignIn(email: string, password: string) {
    try {
      let { data, error }: any = await this.api.signInWithEmail(email, password)
      if (!!error) return { data: null, error }

      if (data?.user?.confirmed_at) {
        this._saveSession(data)
        this._notifyAllSubscribers(AuthChangeEvent.SIGNED_IN)
      }

      return { data, error: null }
    } catch (error) {
      console.log('error_handeEmailSignIn', error)
      return { data: null, error }
    }
  }

  private _handeProviderSignIn(provider: Provider) {
    let url: string = this.api.getUrlForProvider(provider)

    try {
      // try to open on the browser
      if (isBrowser()) {
        window.location.href = url
      }

      return { data: url, error: null }
    } catch (error) {
      // fallback to returning the URL
      if (!!url) return { data: url, error: null }
      else return { data: null, error }
    }
  }

  private _saveSession(session: Session) {
    this.currentSession = session
    this.currentUser = session.user
    let tokenExpirySeconds = session['expires_in']

    if (this.autoRefreshToken && tokenExpirySeconds) {
      setTimeout(this._callRefreshToken, (tokenExpirySeconds - 60) * 1000)
    }

    if (this.persistSession) {
      this._persistSession(this.currentSession, tokenExpirySeconds)
    }
  }

  private _persistSession(currentSession: Session, secondsToExpiry: number) {
    const timeNow = Math.round(Date.now() / 1000)
    const expiresAt = timeNow + secondsToExpiry
    const data = { currentSession, expiresAt }
    isBrowser() && localStorage.setItem(STORAGE_KEY, JSON.stringify(data))
  }

  private _removeSession() {
    this.currentSession = null
    this.currentUser = null
    isBrowser() && localStorage.removeItem(STORAGE_KEY)
  }

  private _recoverSession() {
    const json = isBrowser() && localStorage.getItem(STORAGE_KEY)
    if (json) {
      try {
        const data = JSON.parse(json)
        const { currentSession, expiresAt } = data

        const timeNow = Math.round(Date.now() / 1000)
        if (expiresAt < timeNow) {
          console.log('Saved session has expired.')
          this._removeSession()
        } else if (!currentSession || !currentSession.user) {
          console.log('Current session is missing data.')
          this._removeSession()
        } else {
          this.currentSession = currentSession
          this.currentUser = currentSession.user
          // schedule a refresh 60 seconds before token due to expire
          setTimeout(this._callRefreshToken, (expiresAt - timeNow - 60) * 1000)
        }
      } catch (err) {
        console.error(err)
        return null
      }
    }
    return null
  }

  private async _callRefreshToken() {
    try {
      if (this.currentSession?.refresh_token) {
        let data: any = await this.api.refreshAccessToken(this.currentSession?.refresh_token)

        if (data?.access_token) {
          this.currentSession.access_token = data.body['access_token']
          this.currentSession.refresh_token = data.body['refresh_token']
          let tokenExpirySeconds = data.body['expires_in']

          if (this.autoRefreshToken && tokenExpirySeconds) {
            setTimeout(this._callRefreshToken, (tokenExpirySeconds - 60) * 1000)
          }

          if (this.persistSession && this.currentUser) {
            this._persistSession(this.currentSession, tokenExpirySeconds)
          }
        }
        return { data, error: null }
      }
    } catch (error) {
      return { data: null, error }
    }
  }

  private _notifyAllSubscribers(event: AuthChangeEvent) {
    this.stateChangeEmmitters.forEach((x) => x.callback(event, this.currentSession))
  }
}
