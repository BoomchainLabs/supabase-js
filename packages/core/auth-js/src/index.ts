import Admin from './Admin'
import Client from './Client'

export { Admin, Client }
