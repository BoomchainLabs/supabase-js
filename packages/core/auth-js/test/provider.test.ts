import { GoTrueClient } from '../src/index'
import { Provider } from '../src/lib/types'

const GOTRUE_URL = 'http://localhost:9999'

const auth = new GoTrueClient({
  url: GOTRUE_URL,
  autoRefreshToken: false,
  persistSession: true,
})

test('signIn() with Provider', async () => {
  let { error, data } = await auth.signIn({
    provider: Provider.GOOGLE,
  })
  expect(error).toBeNull()
  expect(data).toMatchSnapshot()
})
