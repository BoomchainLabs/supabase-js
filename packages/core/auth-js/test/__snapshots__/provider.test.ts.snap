// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signIn() with Provider 1`] = `"http://localhost:9999/authorize?provider=google"`;

exports[`signIn() with Provider 2`] = `"google"`;

exports[`signIn() with Provider can append a redirectUrl  1`] = `"http://localhost:9999/authorize?provider=google&redirect_to=https%3A%2F%2Flocalhost%3A9000%2Fwelcome"`;

exports[`signIn() with Provider can append a redirectUrl  2`] = `"google"`;

exports[`signIn() with Provider can append multiple options 1`] = `"http://localhost:9999/authorize?provider=github&redirect_to=https%3A%2F%2Flocalhost%3A9000%2Fwelcome&scopes=repo"`;

exports[`signIn() with Provider can append multiple options 2`] = `"github"`;

exports[`signIn() with Provider can append scopes 1`] = `"http://localhost:9999/authorize?provider=github&scopes=repo"`;

exports[`signIn() with Provider can append scopes 2`] = `"github"`;
